File: 3_dynamic_programming/3_Minimum_Window_Substring.py
"""
LC 76. Minimum Window Substring

Given two strings s and t, return the minimum window in s which will contain all the characters in t. If there is no such window in s that covers all characters in t, return the empty string "".

Note that If there is such a window, it is guaranteed that there will always be only one unique minimum window in s.

Example 1:

Input: s = "ADOBECODEBANC", t = "ABC"
Output: "BANC"
Example 2:

Input: s = "a", t = "a"
Output: "a"

Constraints:

1 <= s.length, t.length <= 105
s and t consist of English letters.

Follow up: Could you find an algorithm that runs in O(n) time?
"""
class Solution:
    def minWindow(self, s: str, t: str) -> str:
        
        if len(s) < len(t):
            return ""
        
        hm = collections.Counter(t)
        comp = len(t)
        minStr = ""
        tail = 0
        
        for idx in range(len(s)):
            if hm[s[idx]] > 0:
                comp -= 1
            
            hm[s[idx]] -= 1
            
            while comp == 0:
                length = idx - tail + 1
                
                if len(minStr) == 0 or len(minStr) > length:
                    minStr = s[tail:idx+1]
                
                hm[s[tail]] += 1
                
                if hm[s[tail]] > 0:
                    comp += 1
                
                tail += 1
                
        return minStr

File: 1. Problems/c. Array/3_Minimum_Window_Substring.cpp
// Minimum Window Substring

/*
Given a string S and a string T, find the minimum window in S which will contain all the characters in T in complexity O(n).

Example:

Input: S = "ADOBECODEBANC", T = "ABC"
Output: "BANC"
Note:

If there is no such window in S that covers all characters in T, return the empty string "".
If there is such window, you are guaranteed that there will always be only one unique minimum window in S.
*/

// my solution - optimal
class Solution {
public:
    string minWindow(string s, string t) {
        vector<int> map(128, 0);
        int counter = t.size(); // substring checker
        int min_head = 0, begin=0, end = 0; // head and tail index
        int d = INT_MAX; // length of substring
        
        // create hash
        for (char a : t)
            map[a]++;
         
        while (end < s.size()) {
            
            if (map[s[end++]]-- > 0)
                counter--;
            
            // when valid
            while (counter == 0) {
                if (end - begin < d) {
                    d = end - begin;
                    min_head = begin;
                }
                
                if (map[s[begin++]]++ == 0)
                    counter++;
            }
        }
        
        return (d == INT_MAX) ? "" : s.substr(min_head, d);
    }
};